### PR TITLE
Milestone 3 friendship.v-2

### PR DESCRIPTION
**Pull request by wintan1418**
implement the functionalities requested for third milestone.

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad**. However, there are a few issues to fix to get approved.

### REQUIRED CHANGES: ⚠️ 

> In this milestone, we implement the association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby). For this reason, we don't need the ruby methods of `pending friend` and `confirm friend`. The SQL queries that you added above still do the magic of adding the second row once a friendship is created. This ensures that the tables of user friendship are optimized.

- [ ] When I comment out the methods `pending friend` and `confirm friend` and try to accept a friendship request, I get the error below.
Screenshot from 2021-04-14 15-10-02
- Give this a fix and ensure that you safely optimize the user friendship table with the creation of the second row. Delete the above methods and use only SQL queries.

- [ ] You currently have only 4 association tests. Add more association and validation tests for your models.
- [ ] Add integration tests for your controllers and views using Rspec and CapyBara gems.


- [ ] This is optional but You will definitely have to implement this change in the next milstone since you have to deploy to heroku. 
> You are currently using sqlite as your db. You are supposed to use Postgresql for both development and production. You might as well start implementing this change now


### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

## Ping me on slack for a chit-chat 🆘  Roy Mukuye Or mention me in this PR

File: app/models/user.rb [hunk 1]
```diff
@@ -9,4 +9,40 @@ class User < ApplicationRecord
...
+  #            Friendship.find_by(user_id: user.id, friend_id: current_user)
+  #   friend.status = true
+  #   friend.save
+  # end
+
+  # def pending_friends
+  #   pending_friendships.map do |f|
+  #     User.find(f.friend_id)
+  #   end
+  # end
```
Comment (cyonii) [inline app/models/user.rb]: Please do not leave commented pieces of code as the make your work messy and unreadable. If you need to see previous changes you should check git histories.

**Review by cyonii — CHANGES_REQUESTED** [review]
### STATUS: CHANGES REQUIRED ♻️✍️

Great effort so far, keep up the energy.

#### TO IMPROVE
- [ ] As the previous reviewer requested, please add more validation and association tests for your models, and feature tests using RSpec and Capybara. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

#### NICE TO HAVE
Note that these suggestions are optional, but I highly recommend that you implement them to improve your project.
- [ ] When a friendship request is accepted, users should no longer see the "confirm", "reject", and "cancel" request buttons, this wouldn't make sense as they are friends already right? Your friendship functionality seems broken.
- [ ] Posts timeline should includes posts written by logged-in user and their friends, right now I can only see posts written by me (the logged-in user), please fix this ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

Please make the requested changes and resubmit for review. Happy coding.

> #### <br>You can ping me on Slack @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) to draw my attention in case I do not respond quickly on here. I'm always available and will be glad to help.<br><br

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad and thanks for implementing the changes requested earlier on. The models tests run just fine 👍🏽**. 
However, there are a few issues to fix to get approved.

### REQUIRED CHANGES: ⚠️ 

- [ ] As I previously stated, add integration tests for your controllers and views using Rspec and CapyBara gems. Use [this article](https://culttt.com/2016/07/06/writing-integration-tests-ruby-rails/) and [this one as well](https://guides.rubyonrails.org/testing.html)

#
- [ ] Since I wasn't able to accept a friendship during the first review, I couldn't recommend some best practices because of that. But I strongly suggest that you implement the NICE TO HAVE changes as stated by @cyonii. 
- Once a friendship is accepted, remove the accept/cancel buttons from users display.
- Also, the timeline of the post should include posts from a user and their friends. It currently displays the users posts only leaving out posts from their friends.

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

## Ping me on slack for a chit-chat 🆘  Roy Mukuye Or mention me in this PR

**Review by NtwaliHeritier — APPROVED** [review]
### **STATUS: APPROVED✔✔✔**

Great work making the required changes👍
You can merge👏

Happy coding💻💻💻

**Pull request merged**